credentials/sts/sts.go

```diff
@@ -0,0 +1,427 @@
+	// https://tools.ietf.org/html/rfc8693#section-3, that indicates the type of
+	// the security token in the "subject_token_path" parameter.
+	SubjectTokenType string // Required.
+
+	// ActorTokenPath is a  security token that represents the identity of the
+	// acting party.
+	ActorTokenPath string // Optional.
+
+	// ActorTokenType is an identifier, as described in
+	// https://tools.ietf.org/html/rfc8693#section-3, that indicates the type of
+	// the the security token in the "actor_token_path" parameter.
+	ActorTokenType string // Optional.
+}
+
+// NewCredentials returns a new PerRPCCredentials implementation, configured
+// using opts, which performs token exchange using STS.
+func NewCredentials(opts Options) (credentials.PerRPCCredentials, error) {
+	if err := validateOptions(opts); err != nil {
+		return nil, err
+	}
+
+	// Load the system roots to validate the certificate presented by the STS
+	// endpoint during the TLS handshake.
+	roots, err := loadSystemCertPool()
+	if err != nil {
+		return nil, err
+	}
+
+	return &callCreds{
+		opts:   opts,
+		client: makeHTTPDoer(roots),
+		bs:     makeBackoffStrategy(),
+	}, nil
+}
+
+// defaultBackoffStrategy returns an exponential backoff strategy based on the
+// default exponential backoff config, but with a smaller BaseDelay (because the
+// former is meant more for connection establishment).
+func defaultBackoffStrategy() backoff.Strategy {
+	config := grpcbackoff.DefaultConfig
+	config.BaseDelay = 100 * time.Millisecond
+	return backoff.Exponential{Config: config}
+}
+
+// httpDoer wraps the single method on the http.Client type that we use. This
+// helps with overriding in unittests.
+type httpDoer interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+func makeHTTPClient(roots *x509.CertPool) httpDoer {
+	return &http.Client{
+		Timeout: stsRequestTimeout,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: roots,
+			},
+		},
+	}
+}
+
+// validateOptions performs the following validation checks on opts:
+// - tokenExchangeServiceURI is not empty
+// - tokenExchangeServiceURI is a valid URI with a http(s) scheme
+// - subjectTokenPath and subjectTokenType are not empty.
+func validateOptions(opts Options) error {
+	if opts.TokenExchangeServiceURI == "" {
+		return errors.New("empty token_exchange_service_uri in options")
+	}
+	u, err := url.Parse(opts.TokenExchangeServiceURI)
+	if err != nil {
+		return err
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return fmt.Errorf("scheme is not supported: %s. Only http(s) is supported", u.Scheme)
+	}
+
+	if opts.SubjectTokenPath == "" {
+		return errors.New("required field SubjectTokenPath is not specified")
+	}
+	if opts.SubjectTokenType == "" {
+		return errors.New("required field SubjectTokenType is not specified")
+	}
+	return nil
+}
+
+// callCreds provides the implementation of call credentials based on an STS
+// token exchange.
+type callCreds struct {
+	opts   Options
```

Comment:
One thought is we can decouple the call creds from the sts implementation. So the call creds is a thin wrapper of the sts implementation. This way, we may be able to move sts to other library (cloud libraries?).

Reply:
I tried doing this. The problem seems to be the following:
* Many of our call creds implementations model themselves after the `oauth2.TokenSource` interface which provides a `Token` method to retrieve the token to use. But this method does not take a context, and therefore it would be hard/ugly for us to enforce a timeout with this approach. There is an open issue for the same: https://github.com/golang/oauth2/issues/262.

And based on an offline discussion with @menghanl, we decided to keep the implementation as is.